### ISSUE TITLE
storage: IngestExternalFiles returns "external files have corrupted keys" on Mac

### ISSUE THREAD

**Issue by jeffrey-xiao** [cockroachdb/cockroach#39696]
Ingestion of SSTs with range deletion tombstones with this [patch](https://github.com/cockroachdb/rocksdb/pull/43) returns "external files have corrupted keys" on Macs, but not on Linux. For example, `TestImportIntoCSV` consistently fails with #39689. Removing the line where we set `read_options` fixes this error.

cc @nvanbenschoten @ajkr @tbg

**Comment by nvanbenschoten** [comment]
`IngestRangeDeletionTombstoneWithGlobalSeqno`, the test added in that patch, passes on Mac.

**Comment by nvanbenschoten** [comment]
I think this is a bug with pinning keys in the `FragmentedRangeTombstoneIterator`. This diff appears to fix the issue:
```
diff --git a/db/range_tombstone_fragmenter.h b/db/range_tombstone_fragmenter.h
index a0b77b677..dda1f303d 100644
--- a/db/range_tombstone_fragmenter.h
+++ b/db/range_tombstone_fragmenter.h
@@ -220,8 +220,7 @@ class FragmentedRangeTombstoneIterator : public InternalIterator {
   };

   void MaybePinKey() const {
-    if (pos_ != tombstones_->end() && seq_pos_ != tombstones_->seq_end() &&
-        (pinned_pos_ != pos_ || pinned_seq_pos_ != seq_pos_)) {
+    if (true) {
       current_start_key_.Set(pos_->start_key, *seq_pos_, kTypeRangeDeletion);
       pinned_pos_ = pos_;
       pinned_seq_pos_ = seq_pos_;
```

**Comment by ajkr** [comment]
I wonder if the `seq_pos_ != tombstones_->seq_end()` is failing only in ingestion preparation because the tombstones all have seqnum zero.

**Comment by nvanbenschoten** [comment]
Good question. It looks like `pinned_pos_ != pos_ || pinned_seq_pos_ != seq_pos_` is evaluting to false.

**Comment by ajkr** [comment]
Huh, could it be because `pinned_pos_` and `pinned_seq_pos_` are uninitialized? I really cannot think of any reason why that would evaluate to false when our sequence of calls is `SeekToFirst()`, `key()`, `Next()`, `key()`, `Next()`, `key()`, ...

If we initialize those in the constructor to the `end()` of their respective containers, any chance it changes the outcome?

**Comment by nvanbenschoten** [comment]
Heh yep, this does the trick:
```
diff --git a/db/range_tombstone_fragmenter.cc b/db/range_tombstone_fragmenter.cc
index e3eb18908..39126737b 100644
--- a/db/range_tombstone_fragmenter.cc
+++ b/db/range_tombstone_fragmenter.cc
@@ -224,7 +224,9 @@ FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
       ucmp_(icmp.user_comparator()),
       tombstones_(tombstones),
       upper_bound_(_upper_bound),
-      lower_bound_(_lower_bound) {
+      lower_bound_(_lower_bound),
+      pinned_pos_(tombstones->end()),
+      pinned_seq_pos_(tombstones->seq_end()) {
   assert(tombstones_ != nullptr);
   Invalidate();
 }
@@ -240,7 +242,9 @@ FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
       tombstones_ref_(tombstones),
       tombstones_(tombstones_ref_.get()),
       upper_bound_(_upper_bound),
-      lower_bound_(_lower_bound) {
+      lower_bound_(_lower_bound),
+      pinned_pos_(tombstones->end()),
+      pinned_seq_pos_(tombstones->seq_end()) {
   assert(tombstones_ != nullptr);
   Invalidate();
 }
```

Did Clang change its treatment of uninitialized member variables across the version we use on Mac and Linux?

Anyway, I'm out tomorrow, so free to turn that into a PR and use it to revert the revert of the reverted revert of Jeffrey's original PR. Let's get that in for good!

**Comment by petermattis** [comment]
> Did Clang change its treatment of uninitialized member variables across the version we use on Mac and Linux?

Probably not as that stuff has been in the C++ language spec for a while. This could be due to differences in memory allocator patterns causing the members to be accidentally initialized to zero on Linux for this test, but full of garbage on Mac OS X. If my memory serves me, it is invalid to expect an STL iterator to be initialized to any value: you have to initialize it explicitly.

**Comment by petermattis** [comment]
If I'm reading this correctly, `vector<T>::const_iterator` is a `const T*`. That means there is no default constructor and thus these members were definitely uninitialized.

Note that a `vector<T>::iterator` is invalidated by any mutation to the vector such as `push_back` or `erase`. I think there is another bug lurking here that we need to invalidate `pinned_pos_` whenever `pos_` is changed and `pinned_seq_pos_` whenever `seq_pos_` is changed. Paging @ajkr for confirmation.

**Comment by petermattis** [comment]
> Note that a vector<T>::iterator is invalidated by any mutation to the vector such as push_back or erase. I think there is another bug lurking here that we need to invalidate {pinned_pos_,pinned_seq_pos_ whenever tombstones_ is changed.

Oh, I don't think `tombstones_` ever changes. So this comment can be ignored.

**Comment by jeffrey-xiao** [comment]
Closed by cockroachdb/rocksdb#44.

**Comment by ajkr** [comment]
> Probably not as that stuff has been in the C++ language spec for a while. This could be due to differences in memory allocator patterns causing the members to be accidentally initialized to zero on Linux for this test, but full of garbage on Mac OS X.

@danhhz was telling me this week about how uninitialized variables can be treated differently by the compiler from an initialized variable in that the memory contents of an uninitialized variable might be ignored when referencing it. Actually that conversation is what led me to this guess - if garbage data were read for these uninitialized iterators, it'd be highly unlikely to make `pinned_pos_ != pos_ || pinned_seq_pos_ != seq_pos_` to be false.

**Comment by ajkr** [comment]
BTW, how about reusing this issue to track upstreaming the fix?

**Comment by jeffrey-xiao** [comment]
Upstream fix is facebook/rocksdb#5720.
